network/onchain-discovery/src/lib.rs

```diff
@@ -64,41 +64,57 @@ pub mod service;
 pub mod types;
 
 /// Query our own storage for the latest discovery set and epoch change proof.
-async fn storage_query_discovery_set(
-    storage_read_client: Arc<dyn StorageRead>,
+fn storage_query_discovery_set(
+    libra_db: Arc<dyn DbReader>,
     req_msg: QueryDiscoverySetRequest,
-) -> Result<(QueryDiscoverySetRequest, QueryDiscoverySetResponseWithEvent)> {
-    let storage_req_msg = Into::<UpdateToLatestLedgerRequest>::into(req_msg.clone());
-
-    let res_msg = storage_read_client
-        .update_to_latest_ledger(
-            storage_req_msg.client_known_version,
-            storage_req_msg.requested_items,
-        )
-        .await
+) -> Result<(QueryDiscoverySetRequest, Box<QueryDiscoverySetResponse>)> {
+    // TODO(philiphayes): how to deal with partial epoch change proof?
+    let (latest_li, epoch_change_proof, accumulator_proof) = libra_db
+        .get_state_proof(req_msg.known_version)
         .with_context(|| {
             format!(
-                "error forwarding discovery query to storage: client version: {}",
-                req_msg.client_known_version,
+                "error getting state proof from storage: request version: {}",
+                req_msg.known_version,
             )
         })?;
 
-    let (response_items, ledger_info_with_sigs, epoch_change_proof, ledger_consistency_proof) =
-        res_msg;
+    // Only return the discovery set account if the requestor is not in the most
+    // recent epoch.
+    let account_state = if epoch_change_proof.ledger_info_with_sigs.is_empty() {
+        None
+    } else {
+        Some(
+            libra_db.get_account_state_with_proof(
+                account_config::discovery_set_address(), req_msg.known_version, latest_li.ledger_info().version()
+            ).with_context(|| {
+                format!("error getting discovery account state with proof from storage: request version: {}, ledger info version: {}",
+                    req_msg.known_version, latest_li.ledger_info().version(),
+                )
+            })?
+        )
+    };
 
-    let res_msg = UpdateToLatestLedgerResponse::new(
-        response_items,
-        ledger_info_with_sigs,
+    let res_msg = Box::new(QueryDiscoverySetResponse {
+        latest_li,
```

Comment:
feels like this would be cleaner to respond None or Some, there doesn't seem to be much value in returning a proof of None

Reply:
i think i will tackle this when i work on supporting partial epoch change proofs, since that's a case i want to make sure we handle properly